src/libutil/tests/characterization.hh

```diff
@@ -0,0 +1,109 @@
+#pragma once
+///@file
+
+#include <gtest/gtest.h>
+
+#include "types.hh"
+
+namespace nix {
+
+/**
+ * The path to the `unit-test-data` directory. See the contributing
+ * guide in the manual for further details.
+ */
+static Path getUnitTestData() {
+    return getEnv("_NIX_TEST_UNIT_DATA").value();
+}
+
+/**
+ * Whether we should update "golden masters" instead of running tests
+ * against them. See the contributing guide in the manual for further
+ * details.
+ */
+static bool testAccept() {
+    return getEnv("_NIX_TEST_ACCEPT") == "1";
+}
+
+/**
+ * Mixin class for writing characterization tests
+ */
+class CharacterizationTest : public virtual ::testing::Test
+{
+protected:
+    /**
+     * While the "golden master" for this characterization test is
+     * located. It should not be shared with any other test.
+     */
+    virtual Path goldenMaster(PathView testStem) const = 0;
+
+public:
+    /**
+     * Golden test for reading
+     *
+     * @param test hook that takes the contents of the file and does the
+     * actual work
+     */
+    void readTest(PathView testStem, auto test)
```

Comment:
```suggestion
    void readTest(PathView testStem, const auto & test)
```

This would be great as it guarantees that `test` is not a function object with state that changes over multiple test executions...

Reply:
I think it is better to keep it `auto` or `auto &&` to encourage inlining.

We can document this should be a "one-shot" lambda which is created and consumed each test run (c.f. Rust's `FnOnce`).

Reply:
We discussed it some more, and it seems like C++ forwarding references are the best choice.